tools/testtool/main.go

```diff
@@ -0,0 +1,203 @@
+		KernelPath:      "/Image",
+		QEMUContainer:   "ghcr.io/hugelgupf/vmtest/qemu:main",
+		QEMUCmd:         "qemu-system-aarch64 -machine virt -cpu max -m 1G -L %s",
+		QEMUPath:        "/zqemu/bin/qemu-system-arm64",
+		BIOSPath:        "/zqemu/pc-bios",
+	},
+}
+
+func defaultConfig() TestEnvConfig {
+	arch := os.Getenv("UROOT_TESTARCH")
+	if c, ok := configs[arch]; ok {
+		return c
+	}
+	if c, ok := configs[runtime.GOARCH]; ok {
+		return c
+	}
+	// On other architectures, user has to provide all values via flags.
+	return TestEnvConfig{}
+}
+
+func run() error {
+	config := defaultConfig()
+	config.RegisterFlags(flag.CommandLine)
+	containerized := flag.Bool("containerized", false, "Run the provided command in a containerized environment instead of your native environment")
+	flag.Parse()
+	if flag.NArg() < 2 {
+		return fmt.Errorf("too few arguments")
+	}
+
+	ctx := context.Background()
+	client, err := dagger.Connect(ctx, dagger.WithLogOutput(os.Stdout))
+	if err != nil {
+		return fmt.Errorf("unable to connect to client: %w", err)
+	}
+	defer client.Close()
+
+	artifacts := client.
+		Container().
+		From(config.QEMUContainer).
+		WithFile(config.KernelPath, client.Container().From(config.KernelContainer).File(config.KernelPath)).
+		Directory("/")
+
+	if *containerized {
+		return runContainerized(ctx, client, artifacts, config.KernelPath, config.QEMUCmd, flag.Args())
+	}
+
+	return runNatively(ctx, artifacts, config.KernelPath, config.QEMUCmd, flag.Args())
+}
+
+func runNatively(
+	ctx context.Context,
+	artifacts *dagger.Directory,
+	kpath, qemuCmd string,
+	args []string,
+) error {
+	tmp, err := os.MkdirTemp(".", "ci-testing")
+	if err != nil {
+		return fmt.Errorf("unable to create tmp dir: %w", err)
+	}
+	defer os.RemoveAll(tmp)
+
+	if ok, err := artifacts.Export(ctx, tmp); !ok || err != nil {
+		return fmt.Errorf("failed artifact export: %w", err)
+	}
+
+	pwd, err := os.Getwd()
+	if err != nil {
+		return fmt.Errorf("could not retrieve working directory: %w", err)
+	}
+
+	tmp, err = filepath.Abs(tmp)
+	if err != nil {
+		return fmt.Errorf("could not retrieve absolute path: %w", err)
+	}
+
+	kpath = filepath.Join(tmp, kpath)
+
+	qemuCmd = fmt.Sprintf(qemuCmd, filepath.Join(tmp, "zqemu", "pc-bios"))
+
+	// Rather than adding the QEMU Cmd to PATH in cmd.Env,
+	// we are doing this because args[0] can be qemu, and if that's the case,
+	// exec.Command does not evaluate the PATH in cmd.Env, but instead the one in the current environment.
+	// The PATH will also be restored after the program exits.
+	p := os.Getenv("PATH")
+	if err := os.Setenv("PATH", fmt.Sprintf("%s:%s", p, filepath.Join(tmp, "zqemu", "bin"))); err != nil {
+		return fmt.Errorf("failed to update PATH: %w", err)
+	}
+	defer os.Setenv("PATH", p)
+
+	cmd := exec.Command(args[0], args[1:]...)
+	cmd.Env = append(
+		os.Environ(),
+		fmt.Sprintf("UROOT_KERNEL=%s", kpath),
+		fmt.Sprintf("UROOT_QEMU=%s", qemuCmd),
+		fmt.Sprintf("UROOT_SOURCE=%s", pwd),
```

Comment:
os.Environ is passed along here. I would not prescribe the values the caller has to use for these last 3 environment variables.

Reply:
But we create a temp dir for the test in which we place the kernel for example so we would still need to set e.g. `UROOT_KERNEL` to point to this path. I opted to only apply this env var to the subprocess in order to not touch the users env here. Same goes for the other ones, they are only set for the subprocess as they got special setup earlier in the function.

Reply:
That only applies to UROOT_KERNEL and UROOT_QEMU, they're the extracted artifacts. All the other env vars - UROOT_SOURCE, UROOT_QEMU_COVERPROFILE, UROOT_QEMU_TIMEOUT_X do not need to be set (and the way you are setting them, the caller cannot set their own!)